prost-build/src/lib_generator.rs

```diff
@@ -0,0 +1,120 @@
+use std::collections::{BTreeMap, BTreeSet};
+use std::fs::read_to_string;
+use std::path::PathBuf;
+
+use crate::{Config, Module};
+
+#[derive(Debug)]
+pub struct Mod {
+    submodules: BTreeMap<String, Mod>,
+    contents: Vec<Module>,
+}
+
+impl Mod {
+    pub fn push(&mut self, module: &Module) {
+        match module.as_slice() {
+            [] => (),
+            [name, left @ ..] => self.add(module.to_owned(), name, left),
+        }
+    }
+
+    fn add(&mut self, module: Module, name: &str, left: &[String]) {
+        let sub = self.submodules.entry(name.to_owned()).or_default();
+        match left {
+            [] => sub.contents.push(module),
+            [name, left @ ..] => sub.add(module, name, left),
+        }
+    }
+}
+
+impl Default for Mod {
+    fn default() -> Self {
+        Self {
+            submodules: BTreeMap::new(),
+            contents: Vec::new(),
+        }
+    }
+}
+
+pub struct LibGenerator<'a> {
+    config: &'a mut Config,
+    depth: u8,
+    buf: &'a mut String,
+}
+
+impl<'a> LibGenerator<'a> {
+    pub fn generate_librs(config: &'a mut Config, mods: &Mod, buf: &'a mut String) {
```

Comment:
i like this a lot! it's similar to how i'm currently using prost, but this is nicer.

what do you think of optionally adding a pointer to the [FileDescriptorSet](https://docs.rs/prost-build/0.7.0/prost_build/struct.Config.html#method.file_descriptor_set_path) in the generated `lib.rs`? Something like 
```rust
pub const FILE_DESCRIPTOR_SET: &[u8] = include_bytes!("./proto_compile/file_descriptor_set.bin");
```

Reply:
Yup, good idea. I actually forgot about this option!

Though it will be a bit more complicated, as outputting the `.bin` is not possible through the protoc plugin interface, as the protocol uses the proto `string` type, which is checked for valid utf8 contents by protoc.

So either the file is written by the plugin itself, which is not how protoc plugins are supposed to work (though it _would_ work), or the contents is outputted to some other `.rs` file, already expanded to some `vec![0x00, 0x43, 0xa3, ..]` expression. That'd be verbose, but it'd work a treat.

I'll try adding the second option.

Reply:
Done in 2a97b84. It adds the `file_descriptor_set` option to the plugin, that can optionally take the name of the generated file (defaults to `file_descriptor_set`).

It adds a `include!("../gen/file_descriptor_set.rs");` to `lib.rs` (or `mod.rs` if not generating a full crate), and the `.rs` file is a vec of all bytes, e.g.

![image](https://user-images.githubusercontent.com/4009336/124786287-e0ac6a80-df47-11eb-9742-34f04d2a4b38.png)